File: app/src/main/java/com/github/benji377/timety/data/FocusMode.kt
package com.github.benji377.timety.data

import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity(tableName = "focus_mode")
data class FocusMode(
    @PrimaryKey(autoGenerate = true) val id: Int = 0,
    val title: String,
    val isCustom: Boolean = true,
    val steps: List<FocusStep>
)

data class FocusStep(
    val durationMins: Int,
    val type: FocusStepType,
    val behavior: FocusStepBehavior = FocusStepBehavior.COUNT_DOWN
)

enum class FocusStepType {
    START, FOCUS, REST, END, LOOP, STOPWATCH, FLEXIBLE
}

enum class FocusStepBehavior {
    COUNT_UP, COUNT_DOWN
}

File: lib/data/focus_mode.dart
import 'dart:convert';

enum FocusStepType { start, focus, rest, end, loop, stopwatch, flexible }

enum FocusStepBehavior { countUp, countDown }

class FocusStep {
  final int durationMins;
  final FocusStepType type;
  final FocusStepBehavior behavior;

  FocusStep({
    required this.durationMins,
    required this.type,
    this.behavior = FocusStepBehavior.countDown,
  });

  Map<String, dynamic> toMap() {
    return {
      'durationMins': durationMins,
      'type': type.index,
      'behavior': behavior.index,
    };
  }

  factory FocusStep.fromMap(Map<String, dynamic> map) {
    return FocusStep(
      durationMins: map['durationMins'],
      type: FocusStepType.values[map['type']],
      behavior: FocusStepBehavior.values[map['behavior']],
    );
  }
}

class FocusMode {
  final int? id;
  final String title;
  final bool isCustom;
  final List<FocusStep> steps;

  FocusMode({
    this.id,
    required this.title,
    this.isCustom = true,
    required this.steps,
  });

  Map<String, dynamic> toMap() {
    return {
      if (id != null) 'id': id,
      'title': title,
      'isCustom': isCustom ? 1 : 0,
      'steps': jsonEncode(steps.map((s) => s.toMap()).toList()),
    };
  }

  factory FocusMode.fromMap(Map<String, dynamic> map) {
    return FocusMode(
      id: map['id'],
      title: map['title'],
      isCustom: (map['isCustom'] ?? 1) == 1,
      steps: map['steps'] != null 
          ? (jsonDecode(map['steps']) as List).map((s) => FocusStep.fromMap(s)).toList()
          : [],
    );
  }
}
